### DISCUSSION TITLE
Match by HTTP Header

### DISCUSSION THREAD

**Discussion by donaldpipowitch** [mswjs/msw#1550] [Q&A] [answered]
### Scope

Adds a new behavior

### Compatibility

- [ ] This is a breaking change

### Feature description

There is a great official Storybook plugin for MSW ([msw-storybook-addon](https://www.npmjs.com/package/msw-storybook-addon)), but it doesn't support sophisticated documentation of multiple stories (which use the same endpoints) within the same page.

This could be fixed on a architecture level (isolating every Story within in an iframe with independent workers), but imho it can be fixed in a simpler way similar to how CSS-in-JS usually fixed "global CSS" (by a proper "scoping" concept).

I'm a new MSW user and if I saw correctly it is currently only possible to [match by method and URL](https://mswjs.io/docs/basics/request-matching). It would actually be nice to match by HTTP Headers as well. It would be pretty easy to append a Story-specific unique HTTP Header to every request. If this would be the case I could easily match against it. With more sophisticated matching options we could support https://github.com/mswjs/msw-storybook-addon/issues/83 in user land.

```tsx
export const Example1: Story = {
  parameters: {
    msw: {
      handlers: [
        rest.get(
          {
            path: '/example',
            headers: {
              'x-msw-mock': 'example-1',
            },
          },
          (req, res, ctx) => {
            return res(ctx.json('Hello'));
          }
        ),
      ],
    },
  },
};

export const Example2: Story = {
  parameters: {
    msw: {
      handlers: [
        rest.get(
          {
            path: '/example',
            headers: {
              'x-msw-mock': 'example-2',
            },
          },
          (req, res, ctx) => {
            return res(ctx.json('World'));
          }
        ),
      ],
    },
  },
};
```

There's definitely room for improving the API (e.g. by providing a `rest` that is already scoped):

```tsx
export const Example1: Story = {
  parameters: {
    msw: {
      scope: 'example-1',
      handlers: ({ rest }) => [
        rest.get('/example', (req, res, ctx) => {
          return res(ctx.json('Hello'));
        }),
      ],
    },
  },
};

export const Example2: Story = {
  parameters: {
    msw: {
      scope: 'example-2',
      handlers: ({ rest }) => [
        rest.get('/example', (req, res, ctx) => {
          return res(ctx.json('World'));
        }),
      ],
    },
  },
};
```

In our application code we would need to access this `scope` (e.g. via a React hook) and append it to all requests (e.g. in an Axios interceptor, if you use Axios).

**Comment by kettanaito** [comment, marked as answer]
Hey, @donaldpipowitch. Thanks for raising this. 

I explain how to implement a similar custom matching functionality [in this discussion](https://github.com/mswjs/msw/discussions/1315#discussioncomment-3108065). 

In your particular case, you can create a custom matcher based on request headers like this:

```js
// byHeaders.js
function byHeaders(expectedHeaders, resolver) {
  return ['*', (req, res, ctx) => {
    const hasMatchingHeaders = Object.entries(expectedHeaders).every(([key, value]) => {
      return req.headers.get(key) === value
    })

    if (hasMatchingHeaders) {
      return resolver(req, res, ctx)
    }

    return req.passthrough()
  }]
}

// Use it in your stories.
import { rest } from 'msw'
import { byHeaders } from './byHeaders'

export const ExampleOne = {
  parameters: {
    msw: {
      handlers: [
        rest.get(...byHeaders({ 'X-Story-Id': '1' }), (req, res, ctx) => {
          return res(ctx.json({ mock: 1 }))
        })
      ]
    }
  }
}

export const ExampleTwo = {
  parameters: {
    msw: {
      handlers: [
        rest.get(...byHeaders({ 'X-Story-Id': '2' }), (req, res, ctx) => {
          return res(ctx.json({ mock: 2 }))
        })
      ]
    }
  }
}
```

MSW API is versatile enough so you could implement various matching algorithms to suit your needs. This doesn't require any changes to the library, and supporting path- and method-based matching is a great default for the majority of use cases.

But from your use case, I would first recommend building that header matching logic straight into the handlers so you don't have to design any custom matchers:

```js
export const ExampleOne = {
  parameters: {
    msw: {
      handlers: [
        rest.get('/resource', (req, res, ctx) => {
          if (req.headers.get('X-Story-Id') === '1') {
            return res(ctx.json({ mock: 1 }))
          }

          return req.passthrough()
        })
      ]
    }
  }
}

export const ExampleTwo = {
  parameters: {
    msw: {
      handlers: [
        rest.get(...byHeaders({ 'X-Story-Id': '2' }), (req, res, ctx) => {
          if (req.headers.get('X-Story-Id') === '2') {
            return res(ctx.json({ mock: 2 }))
          }

          return req.passthrough()
        })
      ]
    }
  }
}
```

**Reply by kettanaito** [reply]
The library is unlikely to ship an API like this because it's too specific. Some may want matchers based on headers, the others may require matching requests based on `mode` or other criteria. All those concerns must be covered by the end consumer, which gives them more control and flexibility over the implementation.

**Reply by donaldpipowitch** [reply]
Thank you for your detailed explanation!
